Add `Window.handle`

damusss opened the pull request:
Window manager information and especially the window handle were only accessible to the display module. I feel like you should know the window handle of any window, so I added the property.  This PR changed so now only the window handle is returned. Check the comments ( [especially this comment](https://github.com/pygame-community/pygame-ce/pull/3219/#issuecomment-2571632176) ) for more.

damusss committed:
Add Window.wm_info

ankith26 reviewed (changes requested):
`SDL_syswm.h` header is removed in SDL3. The way to access the wm info has also changed, and I believe that also changes what values are exported by the SDL side.

I think that going forward any new API added should be SDL3 style, and any thing in the grey area should not be exported at all.

damusss commented:
> `SDL_syswm.h` header is removed in SDL3. The way to access the wm info has also changed, and I believe that also changes what values are exported by the SDL side.
> 
> I think that going forward any new API added should be SDL3 style, and any thing in the grey area should not be exported at all.

Alright, good to know. I'll study the new system that SDL3 uses and adjust the Window code for it. I believe I should keep the SDL2 version (unless it has more dictionary keys than sdl3 will support) so it can be added before pygame 3.

damusss committed:
Use SDL3 Style

damusss commented:
@ankith26 @Starbuck5 
the most recent commit adds a codepath for SDL3. I removed the dictionary keys from the SDL2 codepath that weren't accessible in SDL3. My only limitations:
- I have no idea on how to test the SDL3 codepath
- I got confused sometimes. because I tried to be close to the sdl2 codepath, but sometimes` PyLong_FromLong` was used when the returned value was a pointer, instead of using `PyCapsule_New`. So I'd like to know if what I did is correct, as I assume the checks won't tell me anything

damusss committed:
Update test

ankith26 commented:
Perhaps this PR should wait for the "port window to SDL3" PR to be made and merged first

damusss commented:
> Perhaps this PR should wait for the "port window to SDL3" PR to be made and merged first

I suppose that makes sense. Do you need special competence to pull off one of those PRs, or is it just about following the migration guide (and making different codepaths/creating compat macros)? (if so, I could even try to make it, otherwise I'll wait for you or starbuck). :)

ankith26 commented:
We do test that the patch compiles on SDL3. It can be done locally, but you have to compile SDL3 from source if you are gonna go the WSL way (don't worry, it's not that hard). BTW, we also have SDL3 CI now that can do the testing on windows+mac+linux.

Basically the thing is we are yet to figure how to properly deal with the pixelformat changes uniformly, once that is done we can apply the same strategy everywhere and port the next wave of modules

ankith26 commented:
Well actually, `Window` is not blocked by pixelformat changes, I am working on porting it to SDL3 rn

damusss commented:
Alright then, nice, thanks

Starbuck5 commented:
So even if this code works great on SDL3, why should we expose new API in the SDL2 style?

damusss commented:
> So even if this code works great on SDL3, why should we expose new API in the SDL2 style?

We don't have to expose it in a SDL2 style, but the window manager information is very dependent on the os so how would you provide the info without a dictionary?

Starbuck5 commented:
Having something called "wm info" is clearly an SDL2 construction. Well, looking at the SDL code you have it looks like they expose it now as properties on the window object itself.

Starbuck5 commented:
I have marked as a draft, because with uncertainty about API as well as merge conflicts this is not ready to merge.

damusss commented:
> Having something called "wm info" is clearly an SDL2 construction. Well, looking at the SDL code you have it looks like they expose it now as properties on the window object itself.

That's true, but each property is going to exist or not based on the OS and window manager. they are not generic properties that belong to almost every window like a position and a size. The dictionary reflects how uncertain the attributes are to exist. plus they are all very low lever, so putting them besides common attributes would look "ugly" imo.

damusss committed:
Replace it with Window.handle

damusss committed:
Merge branch 'main' into Window-wm_info

damusss commented:
Ok, this API was kinda ugly so I replaced it with `Window.handle`. it's always an integer, casting pointers to `size_t`. 0 on error. This brings more freedom of Window from display. The rest of the display info can still use the old display function, the handle is the only real thing that needs to be added to Window. New MRE:
```py
import pygame

DISPLAY = True
if DISPLAY:
    screen = pygame.display.set_mode((500, 500))
else:
    win = pygame.Window(size=(500, 500))
    screen = win.get_surface()
clock = pygame.Clock()
while True:
    for e in pygame.event.get():
        if e.type == pygame.QUIT or e.type == pygame.WINDOWCLOSE:
            pygame.quit()
            raise SystemExit
        if e.type == pygame.KEYDOWN:
            if DISPLAY:
                print(pygame.display.get_wm_info()["window"])
            else:
                print(win.handle)
    clock.tick(60)
    if DISPLAY:
        pygame.display.flip()
    else:
        win.flip()

```

damusss committed:
Move docs to stubs ready to fix merge conflicts

damusss committed:
Merge branch 'main' into Window-wm_info

damusss committed:
Fix SDL3 compilation and behavior

damusss commented:
I fixed the merge conflict and I also made sure that Window.handle compiles with SDL3 and returns the proper window handle (I modified window.c to be able to create a window with SDL3 so I could call the property). This is now ready to review.

ankith26 commented:
Is there any other usecase for this apart from diagnostics? We probably discussed this already but just so that it's on github - these handle numbers could potentially be useful for embedding foreign windows in pygame-ce or vice-versa. It is probably going to be very platform/configuration specific though. 
`pygame.display` windows currently support `SDL_WINDOWID` based embedding already but there is no equivalent in the `Window` constructor.

damusss commented:
@ankith26 I would say the usecase is using low level functionality of the OS that can do something if they know the window pointer. If you want, I can rename this to `_handle`, similar to `Surface._pixels_address`. How does that sound?

damusss committed:
Merge branch 'pygame-community:main' into Window-wm_info

MyreMylar reviewed (approved):
This seems to make sense to me based on long ago fiddling with window embedding. I'm happy to allow it and then see if there is demand for any other platform specific window properties with defined use cases as we go along.

The way windows are handled is very platform specific in the details and each platform has it's own selection of identifiers that we are unlikely to be able to abstract perfectly to interface with other library code. You can see SDLs `SDL_GetWindowProperties()` just returns a group of identifiers that differ per platform.

Anyway, approved. 👍

zoldalma999 reviewed (approved):
Added a small comment, but apart from that both API-wise and implementation wise it looks good to me. However I would maybe wait for one or two more approvals, because it is new API.

damusss committed:
Fix versionadded

damusss committed:
Fix property name

damusss commented on src_c/window.c:
```diff
@@ -8,6 +8,10 @@
 
+#if !SDL_VERSION_ATLEAST(3, 0, 0)
+#include <SDL_syswm.h>
+#endif
```
I will fix it, but the first stable release of SDL3 is 3.2.0, meaning SDL 3.1.0 isn't really a thing.

damusss committed:
Fix minor issues

damusss committed:
Add additional subsystem checks

zoldalma999 reviewed (approved):
Still looks good to me (apart from the now outdated versionadded note in the docs).

damusss committed:
Update versionadded tag to 2.5.8

damusss committed:
#if defined -> #ifdef, add #endif comments

oddbookworm reviewed (approved):
LGTM! Assuming @ankith26 is happy with the changes, I would like to get this merged

ankith26 reviewed (approved):
Yeah, I'm fine with this PR. Thanks! 🍰
